File: lib/java/opentoken/src/main/java/com/truveta/opentoken/io/PersonAttributesReader.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io;

import java.util.Iterator;
import java.util.Map;

import com.truveta.opentoken.attributes.Attribute;

/**
 * A generic interface for a streaming person attributes reader.
 */
public interface PersonAttributesReader extends Iterator<Map<Class<? extends Attribute>, String>>, AutoCloseable {

    /**
     * Retrieve the next set of person attributes from an input source.
     * <p>
     * Example person attribute map:
     * <code>
     * {
     *   RecordId: 2ea45fee-90c3-494a-a503-36022c9e1281,
     *   FirstName: John,
     *   LastName: Doe,
     *   Sex: Male,
     *   BirthDate: 01/01/2001,
     *   PostalCode: 54321,
     *   SocialSecurityNumber: 123-45-6789
     * }
     * </code>
     * 
     * @return a person attributes map.
     */
    @Override
    public Map<Class<? extends Attribute>, String> next();
}

File: lib/python/opentoken/src/main/opentoken/io/person_attributes_reader.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from abc import ABC, abstractmethod
from typing import Dict, Type, Iterator
from opentoken.attributes.attribute import Attribute


class PersonAttributesReader(ABC, Iterator[Dict[Type[Attribute], str]]):
    """
    A generic interface for a streaming person attributes reader.
    """

    @abstractmethod
    def __next__(self) -> Dict[Type[Attribute], str]:
        """
        Retrieve the next set of person attributes from an input source.

        Example person attribute map:
        {
            RecordIdAttribute: "2ea45fee-90c3-494a-a503-36022c9e1281",
            FirstNameAttribute: "John",
            LastNameAttribute: "Doe",
            SexAttribute: "Male",
            BirthDateAttribute: "01/01/2001",
            PostalCodeAttribute: "54321",
            SocialSecurityNumberAttribute: "123-45-6789"
        }

        Returns:
            A person attributes map.
        """
        pass

    @abstractmethod
    def __iter__(self):
        """Return the iterator object."""
        return self

    @abstractmethod
    def close(self) -> None:
        """Close the reader and release any resources."""
        pass

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()
